File: samples/microsoft/java/enterprise-agent-tutorial/1-idea-to-prototype/Evaluate.java
import com.azure.ai.projects.AIProjectClient;
import com.azure.ai.projects.AIProjectClientBuilder;
import com.azure.ai.projects.models.*;
import com.azure.identity.DefaultAzureCredentialBuilder;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import io.github.cdimascio.dotenv.Dotenv;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class Evaluate {
    public static void main(String[] args) throws Exception {
        Dotenv dotenv = Dotenv.configure().ignoreIfMissing().load();
        
        String projectEndpoint = dotenv.get("PROJECT_ENDPOINT");
        String modelDeploymentName = dotenv.get("MODEL_DEPLOYMENT_NAME");
        String sharepointResourceName = dotenv.get("SHAREPOINT_RESOURCE_NAME");
        
        AIProjectClient client = new AIProjectClientBuilder()
            .endpoint(projectEndpoint)
            .credential(new DefaultAzureCredentialBuilder().build())
            .buildClient();
        
        System.out.println("🧪 Modern Workplace Assistant Evaluation\n");
        
        List<ToolDefinition> tools = new ArrayList<>();
        
        if (sharepointResourceName != null && !sharepointResourceName.isEmpty()) {
            try {
                ConnectionResponse connection = client.getConnection(sharepointResourceName);
                if (connection != null) {
                    tools.add(new SharePointToolDefinition(connection.getId()));
                    System.out.println("✅ SharePoint configured for evaluation\n");
                }
            } catch (Exception e) {
                System.out.println("⚠️  SharePoint unavailable: " + e.getMessage() + "\n");
            }
        }
        
        String instructions = "You are a Modern Workplace Assistant for Contoso Corporation. " +
            "Answer questions using available tools and provide specific, detailed responses.";
        
        Agent agent = client.createAgent(
            modelDeploymentName,
            new AgentCreationOptions()
                .setName("Evaluation Agent")
                .setInstructions(instructions)
                .setTools(tools)
        );
        
        List<Map<String, String>> questions = new ArrayList<>();
        Gson gson = new Gson();
        
        try (BufferedReader br = Files.newBufferedReader(Path.of("questions.jsonl"))) {
            String line;
            while ((line = br.readLine()) != null) {
                @SuppressWarnings("unchecked")
                Map<String, String> q = gson.fromJson(line, Map.class);
                questions.add(q);
            }
        }
        
        List<Map<String, Object>> results = new ArrayList<>();
        
        System.out.println("Running " + questions.size() + " evaluation questions...\n");
        
        for (int i = 0; i < questions.size(); i++) {
            Map<String, String> q = questions.get(i);
            String question = q.get("question");
            String[] expectedKeywords = q.containsKey("expected_keywords") ? 
                q.get("expected_keywords").split(",") : new String[0];
            
            System.out.println("Question " + (i + 1) + "/" + questions.size() + ": " + question);
            
            AgentThread thread = client.createThread();
            client.createMessage(thread.getId(), MessageRole.USER, question);
            
            ThreadRun run = client.createRun(thread.getId(), agent.getId());
            
            while (run.getStatus() == RunStatus.QUEUED || run.getStatus() == RunStatus.IN_PROGRESS) {
                TimeUnit.MILLISECONDS.sleep(1000);
                run = client.getRun(thread.getId(), run.getId());
            }
            
            String response = "";
            if (run.getStatus() == RunStatus.COMPLETED) {
                List<ThreadMessage> messages = client.listMessages(thread.getId(), 
                    new ListMessagesOptions().setOrder(ListSortOrder.DESCENDING));
                
                for (ThreadMessage message : messages) {
                    if (message.getRole() == MessageRole.ASSISTANT) {
                        for (MessageContent content : message.getContentItems()) {
                            if (content instanceof MessageTextContent) {
                                response = ((MessageTextContent) content).getText();
                                break;
                            }
                        }
                        break;
                    }
                }
            }
            
            boolean passed = response.length() > 50;
            if (expectedKeywords.length > 0) {
                boolean hasKeyword = false;
                for (String keyword : expectedKeywords) {
                    if (response.toLowerCase().contains(keyword.trim().toLowerCase())) {
                        hasKeyword = true;
                        break;
                    }
                }
                passed = passed && hasKeyword;
            }
            
            System.out.println("   Status: " + (passed ? "✅ PASS" : "❌ FAIL"));
            System.out.println("   Response length: " + response.length() + " characters\n");
            
            Map<String, Object> result = new HashMap<>();
            result.put("question", question);
            result.put("response", response);
            result.put("passed", passed);
            result.put("response_length", response.length());
            results.add(result);
            
            client.deleteThread(thread.getId());
        }
        
        client.deleteAgent(agent.getId());
        
        long passedCount = results.stream().filter(r -> (Boolean) r.get("passed")).count();
        long failedCount = results.size() - passedCount;
        
        Map<String, Object> summary = new HashMap<>();
        summary.put("total_questions", questions.size());
        summary.put("passed", passedCount);
        summary.put("failed", failedCount);
        summary.put("results", results);
        
        String json = new GsonBuilder().setPrettyPrinting().create().toJson(summary);
        Files.write(Path.of("evaluation_results.json"), json.getBytes());
        
        System.out.println("📊 Evaluation Complete:");
        System.out.println("   Total: " + questions.size());
        System.out.println("   Passed: " + passedCount);
        System.out.println("   Failed: " + failedCount);
        System.out.println("\n📄 Results saved to evaluation_results.json");
    }
}

File: samples/microsoft/typescript/enterprise-agent-tutorial/1-idea-to-prototype/src/evaluate.ts
import { AIProjectClient } from "@azure/ai-projects";
import { DefaultAzureCredential } from "@azure/identity";
import * as dotenv from "dotenv";
import * as fs from "fs";
import * as readline from "readline";

dotenv.config();

const projectEndpoint = process.env.PROJECT_ENDPOINT!;
const modelDeploymentName = process.env.MODEL_DEPLOYMENT_NAME!;
const sharepointResourceName = process.env.SHAREPOINT_RESOURCE_NAME;
const mcpServerUrl = process.env.MCP_SERVER_URL || "https://learn.microsoft.com/api/mcp";

interface EvaluationQuestion {
    question: string;
    expected_keywords?: string;
}

interface EvaluationResult {
    question: string;
    response: string;
    passed: boolean;
    response_length: number;
}

async function runEvaluation(): Promise<void> {
    console.log("🧪 Modern Workplace Assistant Evaluation\n");

    const credential = new DefaultAzureCredential();
    const client = new AIProjectClient(projectEndpoint, credential);

    let sharepointAvailable = false;
    if (sharepointResourceName) {
        try {
            const connections = await client.connections.list();
            const connectionsList: any[] = [];
            for await (const conn of connections) {
                connectionsList.push(conn);
            }
            sharepointAvailable = connectionsList.some((conn: any) => conn.name === sharepointResourceName);
            if (sharepointAvailable) {
                console.log("✅ SharePoint configured for evaluation\n");
            }
        } catch {
            console.log("⚠️  SharePoint unavailable\n");
        }
    }

    const tools: any[] = [];
    if (sharepointAvailable && sharepointResourceName) {
        tools.push({
            type: "sharepoint_grounding",
            sharepoint_grounding: {
                connection_id: sharepointResourceName
            }
        });
    }

    tools.push({
        type: "mcp",
        mcp: {
            server_url: mcpServerUrl,
            server_name: "microsoft_learn"
        }
    });

    const instructions = "You are a Modern Workplace Assistant for Contoso Corporation. " +
        "Answer questions using available tools and provide specific, detailed responses.";

    const agent = await client.agents.createAgent(modelDeploymentName, {
        name: "Evaluation Agent",
        instructions: instructions,
        tools: tools
    });

    console.log(`✅ Agent created for evaluation: ${agent.id}\n`);

    const fileStream = fs.createReadStream("questions.jsonl");
    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    const questions: EvaluationQuestion[] = [];
    for await (const line of rl) {
        if (line.trim()) {
            questions.push(JSON.parse(line));
        }
    }

    console.log(`Running ${questions.length} evaluation questions...\n`);

    const results: EvaluationResult[] = [];
    
    for (let i = 0; i < questions.length; i++) {
        const q = questions[i];
        console.log(`Question ${i + 1}/${questions.length}: ${q.question}`);

        const thread = await client.agents.threads.create();
        await client.agents.messages.create(thread.id, "user", q.question);
        
        const run = await client.agents.runs.create(thread.id, agent.id);
        
        let runStatus = await client.agents.runs.get(thread.id, run.id);
        while (runStatus.status === "in_progress" || runStatus.status === "queued") {
            await new Promise(resolve => setTimeout(resolve, 1000));
            runStatus = await client.agents.runs.get(thread.id, run.id);
        }

        let response = "";
        if (runStatus.status === "completed") {
            const messages = client.agents.messages.list(thread.id);
            for await (const message of messages) {
                if (message.role === "assistant") {
                    if (message.content && message.content.length > 0) {
                        const content = message.content[0];
                        if (content.type === "text" && "text" in content) {
                            response = content.text.value;
                        }
                    }
                    break;
                }
            }
        }

        let passed = response.length > 50;
        if (q.expected_keywords) {
            const keywords = q.expected_keywords.split(',');
            const hasKeyword = keywords.some(k => 
                response.toLowerCase().includes(k.trim().toLowerCase())
            );
            passed = passed && hasKeyword;
        }
        
        console.log(`   Status: ${passed ? "✅ PASS" : "❌ FAIL"}`);
        console.log(`   Response length: ${response.length} characters\n`);

        results.push({
            question: q.question,
            response: response,
            passed: passed,
            response_length: response.length
        });

        await client.agents.threads.delete(thread.id);
    }

    await client.agents.deleteAgent(agent.id);

    const passedCount = results.filter(r => r.passed).length;
    const failedCount = results.length - passedCount;
    
    const summary = {
        total_questions: questions.length,
        passed: passedCount,
        failed: failedCount,
        results: results
    };

    fs.writeFileSync("evaluation_results.json", JSON.stringify(summary, null, 2));

    console.log("📊 Evaluation Complete:");
    console.log(`   Total: ${questions.length}`);
    console.log(`   Passed: ${passedCount}`);
    console.log(`   Failed: ${failedCount}`);
    console.log("\n📄 Results saved to evaluation_results.json");
}

runEvaluation().catch(console.error);
